gcp_variant_transforms/libs/processed_variant.py

```diff
@@ -0,0 +1,283 @@
+        information used for parsing and splitting INFO fields of thei variant.
+      split_alternate_allele_info_fields (bool): If True, splits fields with
+        field_count='A' (i.e., one value for each alternate) among alternates.
+      annotation_fields (List[str]): If provided, this is the list of
+        INFO field names that store variant annotations. The format of how
+        annotations are stored and their names are extracted from header_fields.
+    """
+    self._header_fields = header_fields
+    self._split_alternate_allele_info_fields = (
+      split_alternate_allele_info_fields)
+    self._annotation_fields = annotation_fields
+
+  def create_processed_variant(self, variant):
+    # type: (vcfio.Variant) -> ProcessedVariant
+    """The main factory method for creating ProcessedVariants.
+
+    Args:
+      variant (:class:`vcfio.Variant`): The raw variant information.
+    """
+    proc_var = ProcessedVariant(variant)
+    proc_var._non_alt_infos = {}
+    annotation_field_set = set(self._annotation_fields or [])
+    for key, variant_info in variant.info.iteritems():
+      # TODO(bashir2): field_count should be removed from VariantInfo and
+      # instead looked up from header_fields.
+      if (self._split_alternate_allele_info_fields and
+              variant_info.field_count == _FIELD_COUNT_ALTERNATE_ALLELE):
+        self._add_per_alt_info(proc_var, key, variant_info.data)
+      elif key in annotation_field_set:
+        self._add_annotation(proc_var, key, variant_info.data)
+      else:
+        proc_var._non_alt_infos[key] = variant_info.data
+    return proc_var
+
+  def _add_per_alt_info(self, proc_var, field_name, variant_info_data):
+    if len(variant_info_data) != len(proc_var._alternate_datas):
+      raise ValueError(
+          'Per alternate INFO field {} does not have same cardinality as '
+          ' number of alternates: {} vs {}'.format(
+              field_name, len(variant_info_data),
+              len(proc_var._alternate_datas)))
+    for alt_index, info in enumerate(variant_info_data):
+      proc_var._alternate_datas[alt_index]._infos[field_name] = info
+
+  def _add_annotation(self, proc_var, field_name, data):
+    # type: (ProcessedVariant, str, List[str]) -> None
+    """Adds an annotation INFO field based on the format in the header.
+
+    Args:
+      proc_var (ProcessedVariant): The object to which the annotations are being
+        added.
+      field_name (str): The name of the annotation field.
+      data(List[str]): The data part of the field separated on comma.
+    """
+    if not self._header_fields.infos.has_key(field_name):
+      raise ValueError('{} INFO not found in the header, variant: {}'.format(
+          field_name, proc_var))
+    header_desc = self._header_fields.infos[field_name].desc
+    annotation_names = extract_annotation_names(header_desc)
+    for annotation_str in data:
+      annotations = _extract_annotation_list_with_alt(annotation_str)
```

Comment:
Given that the first element is always the alternate base, I think it's more readable and a bit more efficient to switch the orders: first convert `data` to be a map of alternate_base -> list of split annotations [1]. Then loop through each alternate_base and add the necessary data. This also avoids having the 1-offset in the code (e.g. 'len - 1' below).

[1] I think it would be even cleaner to do the formatting here as well, so the returned data already contains the annotation names together with the data.